File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.cpp
class Solution {
public:
    int LIS(int ind, int prev_ind, int n, vector<int>& nums, vector<vector<int>> &dp){
        if(ind == n) return 0;
        
        if(dp[ind][prev_ind+1] != -1) return dp[ind][prev_ind+1];

        // not take 
        int len = LIS(ind+1, prev_ind, n, nums, dp);
        // take 
        if(prev_ind == -1 || nums[prev_ind] < nums[ind]){
            len = max(len, 1 + LIS(ind+1, ind, n, nums, dp));
        }

        return dp[ind][prev_ind + 1] = len;
    }

    int lengthOfLIS(vector<int>& nums) {
        int n = nums.size();
        vector<vector<int>> dp(n, vector<int>(n+1, -1));
        return LIS(0, -1, n, nums, dp);
    }
};

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
class Solution:
    def helper(self, ind, prev, nums, dp):
        if ind == len(nums):
            return 0
        
        if dp[ind][prev + 1] != -1:
            return dp[ind][prev + 1]

        # not take case 
        not_take = self.helper(ind + 1, prev, nums, dp)

        # take case 
        take = 0
        if prev == -1 or nums[prev] < nums[ind]:
            take = 1 + self.helper(ind + 1, ind, nums, dp)
        
        dp[ind][prev+1] = max(take, not_take)
        return dp[ind][prev+1]


    def lengthOfLIS(self, nums: List[int]) -> int:
        n = len(nums)
        dp = [[-1 for _ in range(n)] for _ in range(n)]
        return self.helper(0, -1, nums, dp)